src/main.cpp

```diff
@@ -3065,11 +3065,28 @@ static bool ShouldCheckTransactions(const CChainParams& chainparams, const CBloc
 
 bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex,
                   CCoinsViewCache& view, const CChainParams& chainparams,
-                  bool fJustCheck, bool fCheckAuthDataRoot)
+                  bool fJustCheck, CheckAs blockChecks)
 {
     AssertLockHeld(cs_main);
 
-    bool fExpensiveChecks = true;
+    // Disable checking the authDataRoot for block templates and slow block
+    // benchmarks.
+    bool fCheckAuthDataRoot = true;
+    switch (blockChecks) {
+    case CheckAs::Block:
+        break;
+    case CheckAs::BlockTemplate:
+    case CheckAs::SlowBenchmark:
+        fCheckAuthDataRoot = false;
+        break;
+    default:
+        assert(false);
+    }
+
+    // Disable checking proofs and signatures for block templates. This is used
+    // by `TestBlockValidity` to avoid checking them twice for transactions that
+    // were already checked when added to the mempool.
+    bool fExpensiveChecks = blockChecks != CheckAs::BlockTemplate;
```

Comment:
Since we already have the `switch`, it would be somewhat more future-proof to explicitly set `fExpensiveChecks` inside the cases there.